### ISSUE TITLE
Should CABAL_DIR really be dropped in the future?

### ISSUE THREAD

**Issue by danidiaz** [haskell/cabal#7455]
There was a [recent question](https://stackoverflow.com/questions/68079550/clean-installation-of-haskell-onto-a-windows-10-machine) in Stack Overflow, dealing with how to put the files related to Cabal in a non-standard folder. That question has a pretty straightforward solution which consists in setting the [`CABAL_DIR`](https://cabal.readthedocs.io/en/3.4/installing-packages.html?highlight=CABAL_DIR#environment-variables) environment variable. Once set, the configuration, the store and other cabal content will be contained in the folder `CABAL_DIR` points to.

However the user guide [says the following](https://cabal.readthedocs.io/en/3.4/installing-packages.html?highlight=CABAL_DIR#environment-variables):

> The CABAL_DIR might be dropped in the future, when cabal-install starts to use XDG Directory specification.

Following the XDG specification is a good idea, but should it require abandoning `CABAL_DIR`? That would forgo some versatility. 

A good compromise could be following `CABAL_DIR` if defined, and follow the XDG specification otherwise. Alternatively, perhaps the burden of maintaining both options is not worth it?

**Comment by jneira** [comment]
In windows systems we suffer the long standing issue with max path limits. A way to workaround them for cabal has been using `CABAL_DIR` (ghcup added recently support for windows and uses it). So for that case (and i guess many others) i would not want change my entire xdg tree of dirs but only cabal.

**Comment by Mikolaj** [comment]
Is it possible that XDG Directory specification specifies a way to define non-standard directories instead of the usual ones?

**Comment by phadej** [comment]
Don't forget about `CABAL_CONFIG`. IMHO that's  the way you should specify the configuration, `CABAL_DIR` is confusing, as in examples like

```
CABAL_DIR=/home/oleg/.cabal
```

and the `$CABAL_DIR/cabal.config` (`/home/oleg/.cabal/cabal.config) having e.g.

```
store-dir: /cabalstore
```

means that store dir **is** `/cabalstore` and not `$CABAL_DIR/store` (`/home/oleg/.cabal/store`).

---

Therefore IMO, I'd drop `CABAL_DIR` entirely, and use only `CABAL_CONFIG` environment variable  to discover the configuration which default place would be in `XDG_CONFIG_HOME` or looking through `XDG_CONFIG_DIRS` **iff** we would be able to use $xdg_config / $home / ... variables in `cabal.config`, so the config can be shared on multi user machines. See XDG related issues.

**Comment by danidiaz** [comment]
@phadej If I understand correctly, what you mean is that `CABAL_DIR` gives a "false assurance". It seems to promise that all cabal-related files will be contained inside the folder, when there's no garantee of it: the cabal config inside `CABAL_DIR` can specify paths outside `CABAL_DIR` itself. Given that the paths in the config file always win, in the end it's better to just specify `CABAL_CONFIG`.

I see the point. However, in the case of the linked Stack Overflow question, going the `CABAL_CONFIG` route (or using the default config) is a bit more work, because the user had to edit a bunch of unfamiliar fields in the cabal config in order to specify the non-standard paths.

If we wanted to ditch `CABAL_DIR` and still be able to easily specify the storage directories, perhaps we could add an optional `--prefix=<somefolder>` argument to `cabal user-config init`, which would create a starting config having all the storage directories under `<somefolder>`. That way we wouldn't need to edit the config file and modify each individual field. 

(It seems that something a bit like it can already be done with the `--augment=CONFIGLINE` argument. One still needs to mention each separate field, but it could nevertheless be an alternative to `CABAL_DIR` for cases like ghcup's, that @jneira mentioned.)

**Comment by phadej** [comment]
How the initial config is created, or how to modify it isn't in any conflict with removing `CABAL_DIR`.

Note: if user did `cabal update` e.g. once, then the config was already generated, so the absolute default (or previous CABAL_DIR) path is hardcoded into the config. So changing the config will require either deleting it and recreating it (I actually don't know what `cabal user-config update` does, exactrly), or modifying it buy hand or using `user-config`.

Let's separate
- How configuration is read / constructed (`CABAL_CONFIG`, `CABAL_DIR`, `~/.cabal/config` or `~/.config/cabal/config` as XDG recommends) etc
- User interfaces to change and modify these. The `cabal user-config` changes etc.

**Comment by jneira** [comment]
> TL;DR CABAL_DIR and XDG conformance are in conflict. Either we should remove CABAL_DIR concept completely, or close XDG issues with wontfix.

As i see it both are incompatible: XDG would be the default base directory structure for application data, config, etc but users could have legitimate reasons to dont want to use it for some tools, temporary or definitely. I hope the XDG specification does not forbid that :slightly_smiling_face: 

For me `CABAL_DIR` is not misleading: it sets the default base dir for cabal instead a fixed one.  At least is is not more misleading than use the default one (the actual one or the suggested one by XDG) and then change  `CABAL_CONFIG` or change directories  in `~/.cabal/config`. I mean, user is setting on purpose all those things, no?

**Comment by phadej** [comment]
`CABAL_DIR` sets default only when there is no configuration (but there always in practice) and for generating of default configuration. Environment variable is wrong mechanism. `cabal user-config init --cabal-dir=....` is a lot better, and simplifies code pathes everywhere.

**Comment by Mikolaj** [comment]
Related to #680, which is now in active development.

**Comment by hasufell** [comment]
What's the migration path? When the user runs the new cabal that dropped `CABAL_DIR`, then what happens (ghcup itself doesn't set config options and never will)? We print a message?

**Comment by Mikolaj** [comment]
I guess, for a version or two, cabal will print a warning and use the `CABAL_DIR`, whenever cabal uses it currently. Then it will print an error and stop. Then print a warning and ignore `CABAL_DIR` despite it being set. Then just ignore `CABAL_DIR` silently.

**Comment by jneira** [comment]
I would wait to have the nice alternative proposed by phadej: `cabal user-config init --cabal-dir=....`

**Comment by hasufell** [comment]
My opinion is that it's not worth fixing this. It'll cause breakage in CIs in possibly silent ways. The gain for dropping it is marginal.

**Comment by danidiaz** [comment]
Adding `--cabal-dir` to `cabal user-config init` would be a good idea in any case.

**Comment by Mikolaj** [comment]
Absolutely.

**Comment by ivanperez-keera** [comment]
> How the initial config is created, or how to modify it isn't in any conflict with removing CABAL_DIR.

I don't think this is entirely true.

Having a mechanism to specify a location where all files will be located with a simple instruction (`export CABAL_DIR=$PWD/.cabal-sandbox`) enables users who are unfamiliar with Haskell (e.g., those that use eDSLs) get up and running.

It may not look like much, but for people using eDSLs, the cost needs to be tiny or we can't get people to adopt them.

Implementing `cabal user-config init --cabal-dir` *before*  taking away `CABAL_DIR` would provide an interface with the same level of complexity, making the lives of people implementing/using eDSLs much easier.

**Comment by ulysses4ever** [comment]
It seems #7386 followed along the lines of the proposal to preserve `CABAL_DIR` (for now).

**Comment by mouse07410** [comment]
In my humble (blunt, impolite) opinion, it would be pretty stupid to abandon or deprecate `CABAL_DIR`.

**Comment by Mikolaj** [comment]
@mouse07410: do you care to elaborate? Do you agree `CABAL_CONFIG` is a better alternative to `CABAL_DIR`, as outlined above? If not, why?

**Comment by ivanperez-keera** [comment]
@phadej Can you explain why it is crucial or so relevant that the cabal config be shared among users in multi-user machines?

I fail to imagine a realistic and common scenario where that'd play an important role.

**Comment by mouse07410** [comment]
> @mouse07410: do you care to elaborate? Do you agree `CABAL_CONFIG` is a better alternative to `CABAL_DIR`, as outlined above? If not, why?

Unless I misunderstand, `CABAL_CONFIG` would merely point at the config file itself, ditching the user to include more explicit definitions and pointers to subdirectories in it than he might care to. 

`CABAL_DIR`, on the other hand, would imply and include the entire "structure" that will be pointed at by just this one env variable, without the need to modify the config file, or worry about where the stuff should go with the XDG "revolution". 

Therefore, if I had to choose between `CABAL_CONFIG` and `CABAL_DIR` - I would select `CABAL_DIR` with no hesitation.
